src/test/compile-fail/crate-visibility-ambiguity.rs

```diff
@@ -0,0 +1,23 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(crate_in_paths)]
+#![feature(crate_visibility_modifier)]
+
+mod m {
+    pub struct Z;
+    pub struct S1(crate (::m::Z)); // OK
+    pub struct S2(::crate ::m::Z); // OK
+    pub struct S3(crate ::m::Z); //~ ERROR undeclared type or module `crate`
```

Comment:
Huh, this error seems to indicate that it is parsing a path `crate::m::Z` which then fails to resolve... is that correct? If so, that seems sort of wrong, right? [Today at least you get an error](https://play.rust-lang.org/?gist=1015758e53b485dba64b567b06cc29de&version=stable), and if we were going to favor *any* interpretation here, I would expect us to treat `crate` as a visibility modifier (wasn't that the whole point of accepting `::crate` only?).

Reply:
Yes, `crate ::m::Z` is parsed as a single path, I explained this choice in https://github.com/rust-lang/rust/issues/45477#issuecomment-340595412.
FWIW, it's also the most conservative solution (since `crate::anything::else` is an error).

>wasn't that the whole point of accepting ::crate only?

I thought it was for teachability / symmetry with `::other_crate::x::y` (https://github.com/rust-lang/rust/issues/44660#issuecomment-330598910).